src/Pate/Proof.hs

```diff
@@ -83,286 +114,343 @@ data EquivTripleBody sym arch where
-
--- we need to plumb through the initial variables, since they
--- are used when describing the triple that proves this function call
--- is valid
-ppProofFunctionCallSpec ::
-  PA.ValidArch arch =>
-  PT.Sym sym ->
-  PS.SimSpec sym arch (ProofFunctionCall sym arch) ->
-  ProofDoc
-ppProofFunctionCallSpec vsym prfCallSpec =
-  PP.vsep $ 
-    [ "Function pre-domain is satisfied before call:"
-    , PP.indent 4 $ ppPatchPairTarget startPair funPair
-    , PP.indent 4 $ ppEquivTriple vsym prfPreTriple
-    , "Function satisfies post-domain upon return:"
-    , PP.indent 4 $ ppPatchPairReturn funPair
-    , PP.indent 4 $ ppProofBlockSlice vsym $ prfFunBody prfCall
-    ] ++ case prfCall of
-      ProofFunctionCall{} ->
-        [ "Continuing after function return satisfies post-domain for caller."
-        , PP.indent 4 $ ppPatchPairReturn contPair
-        , PP.indent 4 $ ppProofBlockSlice vsym $ prfFunCont prfCall
-        ]
-      ProofTailCall{} -> []
-  where
-    startPair = eqPair $ PS.specBody prfPreTriple
-    funPair = eqPair $ prfTriple $ PS.specBody $ prfFunBody prfCall
-    contPair = eqPair $ prfTriple $ PS.specBody $ prfFunCont prfCall
 
-    prfPreTriple = PS.specMap prfFunPre prfCallSpec
-    prfCall = PS.specBody prfCallSpec
+type family ProofBlock prf :: PT.WhichBinary -> *
+type family ProofRegister prf :: MT.Type -> *
+type family ProofPredicate prf :: *
+type family ProofMemCell prf :: Nat -> *
+type family ProofCounterExample prf :: *
+type family ProofContext prf :: *
+type family ProofBV prf :: Nat -> *
+type family ProofBlockExit prf :: *
+type family ProofMacawValue prf :: MT.Type -> *
 
+class (OrdF (ProofRegister prf),
+       OrdF (ProofMemCell prf)) => IsProof prf
 
+data ProofNodeType =
+    ProofTripleType
+  | ProofFunctionCallType
+  | ProofDomainType
+  | ProofBlockSliceType
+  | ProofStatusType
 
-ppEquivTriple :: PA.ValidArch arch => PT.Sym sym -> EquivTriple sym arch -> ProofDoc
-ppEquivTriple vsym triple =
-  PP.vsep
-    [ "Pre-domain:"
-    , PP.indent 4 $ ppStatePredSpec vsym (PS.specMap eqPreDomain triple)
-    , "Post-domain:"
-    , PP.indent 4 $ ppStatePredSpec vsym (eqPostDomain tripleBody)
-    ]
-  where
-    tripleBody = PS.specBody triple
-
-ppStatePredSpec ::
-  forall sym arch.
-  PA.ValidArch arch =>
-  PT.Sym sym ->
-  PE.StatePredSpec sym arch ->
-  ProofDoc
-ppStatePredSpec vsym@(PT.Sym _ _) stpred =
-  ppRegs <> ppStack <> ppMem
+type ProofDomainType = 'ProofDomainType
+type ProofTripleType = 'ProofTripleType
+type ProofFunctionCallType = 'ProofFunctionCallType
+type ProofBlockSliceType = 'ProofBlockSliceType
+type ProofStatusType = 'ProofStatusType
+
+
+data VerificationStatus ce =
+    Unverified
+  | VerificationSkipped
+  | VerificationSuccess
+  | VerificationFail ce
+  deriving (Functor, Foldable, Traversable)
+
+
+-- | An abstract proof object, representing the overall structure of an equivalence proof.
+data ProofApp prf (node :: ProofNodeType -> *) (tp :: ProofNodeType) where
```

Comment:
It would be really useful to have a full description for each constructor, as well as an overview of the other types to see (e.g., `ProofExpr`) and how they fit together.